### ISSUE TITLE
 conflict of parallel request to docker engine

### ISSUE THREAD

**Issue by xiaochenluo** [moby/moby#38816]
There were two process running. They all used docker python SDK to connect to the local docker engine(18.03). One process would to remove a container which met an error cause the container was running ( this is another issue), almost at the same time, another process would to create a container.  Configs of these two container do not conflict . Then I get an error like this:

```
 Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/docker/api/client.py", line 229, in _raise_for_status
    response.raise_for_status()
  File "/usr/local/lib/python3.5/dist-packages/requests/models.py", line 939, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 409 Client Error: Conflict for url: http://192.168.1.120:2375/v1.35/containers/create
```

During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "worker.py", line 205, in process_job
    **job.get('other', {}) 
  File "/usr/local/lib/python3.5/dist-packages/docker/api/container.py", line 411, in create_container
    return self.create_container_from_config(config, name)
  File "/usr/local/lib/python3.5/dist-packages/docker/api/container.py", line 422, in create_container_from_config
    return self._result(res, True)
  File "/usr/local/lib/python3.5/dist-packages/docker/api/client.py", line 235, in _result
    self._raise_for_status(response)
  File "/usr/local/lib/python3.5/dist-packages/docker/api/client.py", line 231, in _raise_for_status
    raise create_api_error_from_http_exception(e)
  File "/usr/local/lib/python3.5/dist-packages/docker/errors.py", line 31, in create_api_error_from_http_exception
    raise cls(e, response=response, explanation=explanation)
docker.errors.APIError: 409 Client Error: Conflict ("You cannot remove a running container 9c31e308532433a712e9e56792726678ca78ff32e7de116097d27c243ed018a0. Stop the container before attempting removal or force remove")
```

 It is normal to perform these two operations at different times which only got  an 409 error when to  remove a running container. Looks like the docker engine or the API server doesn't support concurrency, does it?

**Comment by cpuguy83** [comment]
The errors in these cases look like:

1) container with the given name already exists
2) a container is running and cannot be removed until you stop it.

In short, this looks to be working as expected.
Can you provide more detail as to why you think there is a problem?

**Comment by xiaochenluo** [comment]
@cpuguy83 
the container name were random named by docker, I didn't name them.
running container cannot remove then gets an error is normal. The  problem is that this affects another container creation at the same time. This has happened 2-3 times which looks like docker processing container not parallel

**Comment by cpuguy83** [comment]
@xiaochenluo Still not much detail to go on. If you have exact repro steps (even if they don't work all the time) it would help for me to visualize what is happening.

However, it sounds like you are having a failure to remove and that is causing the failure to create.
Generally speaking, all docker operations are safe for concurrent access, not that there is no possibility for bugs.

**Comment by xiaochenluo** [comment]
@cpuguy83 I'm sure this concurrent access is not safe because I repeated this. 
I created and started container_1. Then removed container_1 and created container_2 at the same time, I got the same error again. 
To repeat this phenomenon, the remove operation and creation operation must be at the same time. In my code, I use multiprocessing.Barrier to sync in Python3.5.

**Comment by cpuguy83** [comment]
@xiaochenluo Maybe you can layout exact steps with container names and such.

Performing a create and a remove in parallel is a racey operation (who knows which one will be processed first) and there is nothing we can do in the code here to prevent that. It is "safe", however, because this won't cause corrupted data in the engine.

**Comment by xiaochenluo** [comment]
@cpuguy83
I had misunderstood the meaning of ‘safe'  : ) 
This situation is not common in our project, and exception catching is used to handling it currently. Swarm or k8s is being considered for future concurrency. Thank you for your reply.

**Comment by thaJeztah** [comment]
Let me close this ticket, as it doesn't look there's a bug at hand, but feel free to continue the conversation
